### ISSUE TITLE
v2.0.0-alpha-2: Support for ES3/5 'modules' (implicit strict mode) is lost

### ISSUE THREAD

**Issue by nre** [eslint/eslint#4832]
`ecmaFeatures.modules = true` in v1.10.3 supports ES3/5 'modules' as well as ES6 modules, whereas `parserOptions.sourceType = "module"` in v2.0.0-alpha-2 only supports ES6 modules (because it silently overrides `parserOptions.ecmaVersion`: https://github.com/eslint/eslint/blob/master/lib/eslint.js#L418).

An ES3/5 'module' may be *implicitly* in strict mode. For example, this file:
```js
// myModule.js
exports.foo = "bar";
```
is implicitly in strict mode because it will be bundled up into something like this:
```js
// bundle.js
loadModules({
    myModule: function (require, exports, module) {
        "use strict";
        // myModule.js
        exports.foo = "bar";
    },
});
```

Support for ES3/5 'modules' and implicit strict mode can be tested with the following files, both of which contain an error:
```js
// es6.js
const ES6 = true;
```
```js
// invalidThis.js
function invalidThis() {
    return this;
}
```

v1.10.3 supports ES3/5 'modules' because it finds both errors:
```json
{
    "ecmaFeatures": {
        "modules": true
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
```
es6.js
  2:2  error  Parsing error: Unexpected token const

invalidThis.js
  3:9  error  Unexpected `this`  no-invalid-this

✖ 2 problems (2 errors, 0 warnings)
```

v2.0.0-alpha-2 does not support ES3/5 'modules' because it can find either the first error:
```json
{
    "parserOptions": {
        "ecmaVersion": 5
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
```
es6.js
  2:2  error  Parsing error: The keyword 'const' is reserved

✖ 1 problem (1 error, 0 warnings)
```
or the second error (but not both):
```json
{
    "parserOptions": {
        "ecmaVersion": 5,
        "sourceType": "module"
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
```
invalidThis.js
  3:9  error  Unexpected `this`  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

Although it is correct that `parserOptions.sourceType = "module"` should imply ES6 (although perhaps not silently), I hope that support for ES3/5 'modules' and implicit strict mode can be restored to v2 in some way.

**Comment by nre** [comment]
I had assumed that all JS bundlers started each module with `"use strict"` by default, but apparently some people are still running/writing non-strict-mode JS!

Here is a list of tools to show that I'm not the only one who writes their ES3/5 'modules' with implicit strict mode, for browsers as well as Node:
- https://www.npmjs.com/package/strictify
- https://www.npmjs.com/package/esformatter-use-strict
- https://www.npmjs.com/package/use-strict
- https://www.npmjs.com/package/strict-mode
- https://www.npmjs.com/package/stricterify
- https://www.npmjs.com/package/strict-mode-ify

**Comment by nzakas** [comment]
Using ES6 modules as a proxy to say "we have implicit strict mode" is not a good solution because modules are not simply implicit strict mode, they imply a lot of different things. 

I'm very unfamiliar with this use case. Are you saying some custom module systems insert strict mode when compiling?

**Comment by michaelficarra** [comment]
@nzakas The only things I can think of:

0. Strict
0. Allow import/export statements
0. Top level declarations are not globals
0. `async` is a keyword

**Comment by nzakas** [comment]
`await` is reserved as well, correct?

**Comment by michaelficarra** [comment]
Oops, that's what I meant. Not `async`.

**Comment by nre** [comment]
The use case is laziness. `"use strict"` in every module seems redundant when I only ever read, write and run strict-mode JS. And the job of build tools is to do these repetitive tasks!

I understand now that the old `ecmaFeatures.modules` option referred to ES6 modules. But the docs said 'enable modules and global strict mode' and I took this to mean what I was already doing: CommonJS modules and implicit strict mode.

Looking at 1.10.3, I see that setting `ecmaFeatures.modules` also set `ecmaVersion` to 6. But `ecmaVersion` was only passed to escope, not to the parser or anyone else. The result was like ES5 CommonJS modules and implicit strict mode.

I've looked for a way to give eslint an 'implicit strict mode' option, but I think it isn't possible. I can see how to fool escope, but unfortunately not the parser. If anyone knows a way then I'd be grateful to hear it.

Otherwise this non-issue can be closed.

**Comment by nzakas** [comment]
It is possible to add an option, but it's a nontrivial amount of work as it would need to change Espree, escope, and ESLint together. 

Do you have other options on your end?

**Comment by nre** [comment]
My options are:

1. Use v2 and get used to having `"use strict";` in my modules. (Not a big deal.)
2. Use v2 and have my ES3/5 modules linted as ES6. (Not great, but I know the difference between ES3/5 and ES6.)
3. Stay on v1 until I start using ES6. (I'm waiting for good browser support first.)

JSHint has a closed issue for `strict: implied` (jshint/jshint#924). You can see that there's moderate demand, but not huge. The feature is supposed to be in v2.9 but that's still in pre-release, so it's not like users who migrate to eslint will be disappointed. (Although I was quite happy when I thought eslint did have the feature!)

I can fool escope without changing it:
```js
var useStrictAst = espree.parse('"use strict";');
var useStrictStatement = useStrictAst.body[0];

ast.body.splice(0, 0, useStrictStatement);
var scopeManager = escope.analyze(ast, options);
ast.body.splice(0, 1);
```
Is this good enough for eslint?

The parser is difficult because it takes a string. A directive is ignored if it is added to the end of the string. Adding a directive to the start means all the start/end character offsets need to be fixed. (Does eslint use these, or only the loc objects? Would only the column numbers of line 1 need to be fixed?) I think the only clean solution is for parsers to expose their `isStrict` initial value.

I agree that this lazybones feature doesn't warrant non-trivial hacking.

**Comment by nzakas** [comment]
Unfortunately that's not a feasible solution. You need to know that it's strict mode during the parsing phase to catch strict mode errors, just tricking escope doesn't result in catching these errors. 

If someone wants to take a stab at this, I envision the steps to be:

1. Add ecmaFeatures.impliedStrict to Espree
1. Add a similar flag to escope
1. Update ESLint to use them and update docs

**Comment by nre** [comment]
TODO
- Add `impliedStrict` to Espree eslint/espree#227
  - [x] PR eslint/espree#228
  - [x] Merged
  - [ ] Released
- Add `impliedStrict` to Escope estools/escope#92
  - [x] PR estools/escope#93
  - [ ] Merged
  - [ ] Released
- Add `impliedStrict` to ESLint eslint/eslint#4832
  - [ ] Tidy up 'strict' rule eslint/eslint#4936
  - [ ] PR

**Comment by nre** [comment]
@nzakas

The `impliedStrict` option has been merged in Escope, but I think it might be weeks/months until it is released. (I have requested a new release: estools/escope#98.) That is the only issue blocking the `impliedStrict` option in ESLint.

Would you reconsider permitting the [AST-mutation Escope workaround](https://github.com/eslint/eslint/issues/4832#issuecomment-168335647)? It would look like this:

```js
/*
 * Temporary workaround until Escope supports the `impliedStrict`
 * option.
 */
shouldWorkAroundEscopeImpliedStrict = ecmaFeatures.impliedStrict;

if (shouldWorkAroundEscopeImpliedStrict) {
    impliedStrictAst = parse("'use strict';", {parser: config.parser});
    impliedStrictDirective = impliedStrictAst.body[0];
    ast.body.unshift(impliedStrictDirective);
}

// gather data that may be needed by the rules
scopeManager = escope.analyze(ast, {
    ignoreEval: true,
    nodejsScope: ecmaFeatures.globalReturn,
    // impliedStrict: ecmaFeatures.impliedStrict,
    ecmaVersion: ecmaVersion,
    sourceType: currentConfig.parserOptions.sourceType || "script"
});
currentScopes = scopeManager.scopes;

if (shouldWorkAroundEscopeImpliedStrict) {
    ast.body.shift();
}
```

I'm not asking this for myself (I'm using my patched versions of Escope and ESLint) but so that the feature is available to anyone who might want it. (JSHint has the feature in its latest stable release.)

**Comment by nzakas** [comment]
Why would there be such a delay for escope?

**Comment by michaelficarra** [comment]
@nre I've published escope 3.4.0.

**Comment by nzakas** [comment]
Thanks @michaelficarra!

**Comment by shaunstanislaus** [comment]
```
SublimeLinter: eslint: start.js ['/usr/local/bin/eslint', '--format', 'compact', '--stdin', '--stdin-filename', '__RELATIVE_TO_FOLDER__'] 
SublimeLinter: eslint output:
es6-boilerplate/server/scripts/start.js: line 3, col 2, Error - Parsing error: The keyword 'const' is reserved

1 problem 
```

'use strict';

```javascript
const args = require('optimist').argv;
const app  = require('../index');

app.setPort(typeof args.p !== 'undefined' ? args.p : args.port);
app.start();

```

**Comment by silvenon** [comment]
@shaunstanislaus this is a long shot, but are you getting that error from AtomLinter/linter-eslint or ESLint itself?

**Comment by shaunstanislaus** [comment]
@silvenon eslint itself.

**Comment by sgnl** [comment]
I was about to post that I'm experiencing this issue but after restarted ST3 about 4 times the error is gone...  ¯\_(ツ)_/¯
